src/main/java/com/brew/oauth20/server/controller/AuthorizeController.java

```diff
@@ -19,67 +23,104 @@ public class AuthorizeController {
+                               AuthorizeTypeProviderFactory authorizeTypeProviderFactory,
+                               AuthorizationCodeService authorizationCodeService,
+                               TokenGrantProviderFactory tokenGrantProviderFactory) {
         this.userCookieService = userCookieService;
         this.authorizeTypeProviderFactory = authorizeTypeProviderFactory;
+        this.tokenGrantProviderFactory = tokenGrantProviderFactory;
         this.authorizationCodeService = authorizationCodeService;
         this.userIdCookieKey = "SESSION_ID";
         this.locationHeaderKey = "Location";
+        this.authorizationHeaderKey = "Authorization";
     }
 
     @GetMapping(value = "/oauth/authorize")
-    public ResponseEntity<String> get(@Valid @ModelAttribute("authorizeRequest") AuthorizeRequestModel authorizeRequest, BindingResult validationResult, HttpServletRequest request) {
+    public ResponseEntity<String> authorizeGet(
+            @Valid @ModelAttribute("authorizeRequest") AuthorizeRequestModel authorizeRequest,
+            BindingResult validationResult,
+            HttpServletRequest request) {
         return authorize(authorizeRequest, validationResult, request);
     }
 
-
     @PostMapping(value = "/oauth/authorize")
-    public ResponseEntity<String> post(@Valid @RequestBody AuthorizeRequestModel authorizeRequest, BindingResult validationResult, HttpServletRequest request) {
+    public ResponseEntity<String> authorizePost(@Valid @RequestBody AuthorizeRequestModel authorizeRequest,
+                                                BindingResult validationResult,
+                                                HttpServletRequest request) {
         return authorize(authorizeRequest, validationResult, request);
     }
 
-    private ResponseEntity<String> authorize(AuthorizeRequestModel authorizeRequest, BindingResult validationResult, HttpServletRequest request) {
+    @PostMapping(value = "/oauth/token")
+    public ResponseEntity<TokenResultModel> tokenPost(@Valid @RequestBody TokenRequestModel tokenRequestModel,
+                                                      BindingResult validationResult,
+                                                      HttpServletRequest request) {
+        if (validationResult.hasErrors()) {
+            return new ResponseEntity<>(new TokenResultModel(null, "invalid_request"), HttpStatus.BAD_REQUEST);
+        }
+
+        var authorizationHeaderValue = request.getHeader(authorizationHeaderKey);
+
+        var tokenGrantProvider = tokenGrantProviderFactory
+                .getService(GrantType.fromValue(tokenRequestModel.grant_type));
+
+        var tokenResponse = tokenGrantProvider.generateToken(authorizationHeaderValue, tokenRequestModel);
+
+        HttpStatus httpStatus = tokenResponse.getError() == null ? HttpStatus.OK : HttpStatus.BAD_REQUEST;
+
+        return new ResponseEntity<>(tokenResponse, httpStatus);
```

Comment:
Hata varsa BadRequest'in içinde error stringi dönelim sadece null token değerine gerek yok. Hata yoksa da error olmadan sadece Token değerini dönelim.

Reply:
bunu anlamadım. bir nesne belirleyip onu dönüyoruz o nesnede TokenResultModel. içinde ya token nesnesi oluyor yada hata